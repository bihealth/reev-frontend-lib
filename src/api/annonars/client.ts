import { chunks } from '@reactgular/chunks'

import type { LinearStrucvar, Seqvar } from '../../lib/genomicVars'
import { urlConfig } from '../../lib/urlConfig'
import { ClinvarPerGeneRecord } from '../../pbs/annonars/clinvar/per_gene'
import { Record as GeneInfoRecord } from '../../pbs/annonars/genes/base'
import {
  ClinvarSvQueryResponse,
  GeneInfoResult,
  GeneSearchResponse,
  SeqvarInfoResponse
} from './types'

/**
 * Client for accessing annonars REST API.
 */
export class AnnonarsClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws Error if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlAnnonars !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlAnnonars
    } else {
      throw new Error('Configuration error: API base URL not configured')
    }
  }

  /**
   * Fetch gene information via annonars REST API.
   *
   * @param hgncId HGNC ID, e.g., `"HGNC:26467"`.
   */
  async fetchGeneInfo(hgncId: string): Promise<GeneInfoResult> {
    const response = await fetch(`${this.apiBaseUrl}/genes/info?hgnc_id=${hgncId}`, {
      method: 'GET'
    })
    return await response.json()
  }

  /**
   * Fetch variant information via annonars and mehari REST APIs.
   *
   * @param seqvar The variant to retrieve the information for.
   */
  async fetchVariantInfo(seqvar: Seqvar): Promise<SeqvarInfoResponse> {
    const { genomeBuild, chrom, pos, del, ins } = seqvar
    let chromosome = chrom.replace('chr', '')
    if (genomeBuild !== 'grch37') {
      chromosome = `chr${chrom}`
    }

    const url =
      `${this.apiBaseUrl}/annos/variant?genome_release=${genomeBuild}&` +
      `chromosome=${chromosome}&pos=${pos}&reference=${del}&` +
      `alternative=${ins}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch variant info: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return SeqvarInfoResponse.fromJson(responseJson)
  }

  /**
   * Fetch per-gene ClinVar information via annonars REST API.
   *
   * @param hgncId
   * @returns
   */
  async fetchGeneClinvarInfo(hgncId: string): Promise<ClinvarPerGeneRecord> {
    const response = await fetch(`${this.apiBaseUrl}/genes/clinvar?hgnc_id=${hgncId}`, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch gene clinvar info: ${response.statusText}`)
    }
    return await response.json()
  }

  /**
   * Search for genes, e.g., by symbol, via annonars REST API.
   *
   * @param query Query string to search for.
   * @returns Promise with gene search response.
   * @throws Error if the request fails.
   */
  async fetchGenes(query: string): Promise<GeneSearchResponse> {
    const response = await fetch(`${this.apiBaseUrl}/genes/search?q=${query}`, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch genes: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return GeneSearchResponse.fromJson(responseJson)
  }

  /**
   * Fetch gene informations via annonars REST API.
   *
   * @param hgncIds Array of HGNC IDs to use, e.g., `["HGNC:26467"]`.
   * @param chunkSize How many IDs to send in one request.
   * @returns Promise with an array of gene information objects.
   */
  async fetchGeneInfos(hgncIds: string[], chunkSize?: number): Promise<GeneInfoRecord[]> {
    const hgncIdChunks = chunks(hgncIds, chunkSize ?? 10)

    const promises = hgncIdChunks.map(async (chunk) => {
      const url = `${this.apiBaseUrl}/genes/info?hgnc_id=${chunk.join(',')}`

      const response = await fetch(url, {
        method: 'GET'
      })
      if (!response.ok) {
        throw new Error(`failed to fetch gene infos: ${response.statusText}`)
      }
      return response
    })

    const responses = await Promise.all(promises)
    const resultJsons = await Promise.all(responses.map((response) => response.json()))

    const result: GeneInfoRecord[] = []
    resultJsons.forEach((chunk: any) => {
      for (const value of Object.values(chunk.genes)) {
        // @ts-ignore
        result.push(GeneInfoRecord.fromJson(value as JsonValue))
      }
    })
    return result
  }

  /**
   * Fetch overlapping ClinVar strucvars via annonars REST API.
   */
  async fetchClinvarStrucvars(
    strucvar: LinearStrucvar,
    pageSize: number = 1000,
    minOverlap: number = 0.1
  ): Promise<ClinvarSvQueryResponse> {
    const { genomeBuild, chrom, start, stop } = strucvar
    const url =
      `${this.apiBaseUrl}/clinvar-sv/query?genomeRelease=${genomeBuild}&` +
      `chromosome=${chrom}&start=${start}&stop=${stop}&pageSize=${pageSize}&` +
      `minOverlap=${minOverlap}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new Error(`failed to fetch clinvar strucvars: ${response.statusText}`)
    }
    const responseJson = await response.json()
    return ClinvarSvQueryResponse.fromJson(responseJson)
  }
}
