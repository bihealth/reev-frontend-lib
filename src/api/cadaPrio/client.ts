import { urlConfig } from '../../lib/urlConfig'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { Response } from './types'

/**
 * Client for the CADA-Prio API.
 */
export class CadaPrioClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws ConfigError if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlCadaPrio !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlCadaPrio
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Predict similarity of genes with list of HPO terms.
   *
   * @param hpoTerms HPO term IDs (e.g. `HP:0000001`)
   * @param geneSymbols Gene symbols (e.g. `BRCA1`)
   * @returns Promise with response of impact prediction.
   * @throws StatusCodeNotOk if the request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async predictGeneImpact(hpoTerms: string[], geneSymbols?: string[]): Promise<Response> {
    const geneSuffix = geneSymbols ? `&gene_symbols=${geneSymbols.join(',')}` : ''
    const url = `${this.apiBaseUrl}/api/v1/predict?hpo_terms=${hpoTerms.join(',')}${geneSuffix}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(`failed to fetch gene impact prediction: ${response.statusText}`)
    }
    try {
      const responseJson = await response.json()
      return Response.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(`failed to parse gene impact prediction response: ${e}`)
    }
  }
}
