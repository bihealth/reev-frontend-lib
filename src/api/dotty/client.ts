import { urlConfig } from '../../lib/urlConfig'
import { SpdiResult, TranscriptResult } from './types'

export class DottyClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws Error if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlDotty !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlDotty
    } else {
      throw new Error('Configuration error: API base URL not configured')
    }
  }

  /**
   * Convert the given query to SPDI notation.
   *
   * @param q Query to conert.
   * @param assembly Assembly to use.
   * @returns Promise with the response form dotty or null if the request failed.
   */
  async toSpdi(q: string, assembly: 'GRCh37' | 'GRCh38' = 'GRCh38'): Promise<SpdiResult | null> {
    const escapedQ = encodeURIComponent(q)
    const url = `${this.apiBaseUrl}/api/v1/to-spdi?q=${escapedQ}&assembly=${assembly}`
    const response = await fetch(url, {
      method: 'GET'
    })
    if (response.status == 200) {
      const responseJson = await response.json()
      return SpdiResult.fromJson(responseJson)
    } else {
      return null
    }
  }

  /**
   * Fetch transcripts for the given HGNC ID.
   *
   * @param hgncId HGNC ID to fetch transcripts for.
   * @param assembly Assembly to use.
   * @returns Promise with the response form dotty or null if the request failed.
   */
  async fetchTranscripts(
    hgncId: string,
    assembly: 'GRCh37' | 'GRCh38' = 'GRCh38'
  ): Promise<TranscriptResult | null> {
    const url = `${this.apiBaseUrl}/api/v1/find-transcripts?hgnc_id=${hgncId}&assembly=${assembly}`
    const response = await fetch(url, {
      method: 'GET'
    })
    if (response.status == 200) {
      const responseJson = await response.json()
      return TranscriptResult.fromJson(responseJson)
    } else {
      return null
    }
  }
}
