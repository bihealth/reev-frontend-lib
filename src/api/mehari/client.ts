import type { Seqvar, Strucvar } from '../../lib/genomicVars'
import { urlConfig } from '../../lib/urlConfig'
import { GeneTranscriptsResponse } from '../../pbs/mehari/server'
import { GenomeBuild } from '../../pbs/mehari/txs'
import { ConfigError, InvalidResponseContent, StatusCodeNotOk } from '../common'
import { SeqvarResult, StrucvarResult } from './types'

/**
 * Client for Mehari API requests.
 */
export class MehariClient {
  private apiBaseUrl: string

  /**
   * @param apiBaseUrl
   *            API base to the backend, excluding trailing `/`.
   *            The default is declared in '@/lib/urlConfig`.
   * @throws ConfigError if the API base URL is not configured.
   */
  constructor(apiBaseUrl?: string) {
    if (apiBaseUrl !== undefined || urlConfig.baseUrlMehari !== undefined) {
      // @ts-ignore
      this.apiBaseUrl = apiBaseUrl ?? urlConfig.baseUrlMehari
    } else {
      throw new ConfigError('Configuration error: API base URL not configured')
    }
  }

  /**
   * Retrieve consequences for sequence variants.
   *
   * @param seqvar Sequence variant to retrieve consequences for.
   * @param hgncId HGNC ID of gene to restrict results to.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveSeqvarsCsq(seqvar: Seqvar, hgncId?: string): Promise<SeqvarResult> {
    const { genomeBuild, chrom, pos, del, ins } = seqvar
    const hgncSuffix = hgncId ? `&hgnc_id=${hgncId}` : ''
    const url =
      `${this.apiBaseUrl}/api/v1/seqvars/csq?genome_release=${genomeBuild}&` +
      `chromosome=${chrom}&position=${pos}&reference=${del}&` +
      `alternative=${ins}${hgncSuffix}`

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(
        `Failed to fetch sequence variant consequences: ${response.statusText}`
      )
    }
    try {
      const responseJson = await response.json()
      return SeqvarResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(
        `Failed to parse sequence variant consequences response: ${e}`
      )
    }
  }

  /**
   * Retrieve consequences for structural variants.
   *
   * @param strucvar Structural variant to retrieve consequences for.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveStrucvarsCsq(strucvar: Strucvar): Promise<StrucvarResult> {
    let url: string
    const { svType, genomeBuild, chrom, start } = strucvar
    if (svType === 'BND' || svType === 'INS') {
      url =
        `${this.apiBaseUrl}/api/v1/strucvars/csq?genome_release=${genomeBuild}&` +
        `chromosome=${chrom}&start=${start}&stop=${start + 1}&sv_type=${svType}`
    } else {
      const { stop } = strucvar
      url =
        `${this.apiBaseUrl}/api/v1/strucvars/csq?genome_release=${genomeBuild}&` +
        `chromosome=${chrom}&start=${start}&stop=${stop}&sv_type=${svType}`
    }

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(
        `Failed to fetch structural variant consequences: ${response.statusText}`
      )
    }
    try {
      const responseJson = await response.json()
      return StrucvarResult.fromJson(responseJson)
    } catch (e) {
      throw new InvalidResponseContent(
        `Failed to parse structural variant consequences response: ${e}`
      )
    }
  }

  /**
   * Retrieve transcripts given a HGNC ID.
   *
   * @param hgncId HGNC ID of gene to retrieve transcripts for.
   * @param genomeBuild Genome build to restrict results to.
   * @param pageSize Number of results to return per page.
   * @param nextPageToken Token to retrieve the next page of results.
   * @returns Promise with the response from the API.
   * @throws StatusCodeNotOk if the API request fails.
   * @throws InvalidResponseContent if the response is not valid JSON.
   */
  async retrieveGeneTranscripts(
    hgncId: string,
    genomeBuild: GenomeBuild,
    pageSize: number = 1000,
    nextPageToken?: string
  ): Promise<GeneTranscriptsResponse> {
    const urlGenomeBuild = genomeBuild === GenomeBuild.GENOME_BUILD_GRCH37 ? 'grch37' : 'grch38'
    const url =
      `${this.apiBaseUrl}/api/v1/genes/transcripts?hgnc_id=${hgncId}&` +
      `genome_build=${urlGenomeBuild}&page_size=${pageSize}` +
      (nextPageToken ? `&next_page_token=${nextPageToken}` : '')

    const response = await fetch(url, {
      method: 'GET'
    })
    if (!response.ok) {
      throw new StatusCodeNotOk(`Failed to fetch transcripts: ${response.statusText}`)
    }
    try {
      const responseJson = await response.json()
      return GeneTranscriptsResponse.fromJson(responseJson, { ignoreUnknownFields: true })
    } catch (e) {
      throw new InvalidResponseContent(`Failed to parse transcript response: ${e}`)
    }
  }
}
