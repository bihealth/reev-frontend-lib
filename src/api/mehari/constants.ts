/** Mapping from transcript effect to label. */
export const TX_EFFECT_LABELS: { [key: string]: string } = {
  transcript_variant: 'whole tx',
  exon_variant: 'exonic',
  splice_region_variant: 'splicing',
  intron_variant: 'intronic',
  upstream_variant: 'upstream',
  downstream_variant: 'downstream',
  other: 'other'
}
