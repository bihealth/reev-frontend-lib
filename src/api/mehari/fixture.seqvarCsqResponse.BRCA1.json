{
  "version": { "tx_db": "0.5.0", "mehari": "0.7.0" },
  "query": {
    "genome_release": "grch37",
    "chromosome": "chr17",
    "position": 41197751,
    "reference": "G",
    "alternative": "T",
    "hgnc_id": null
  },
  "result": [
    {
      "consequences": ["MissenseVariant"],
      "feature_tag": ["ManeSelect"],
      "putative_impact": "Moderate",
      "gene_symbol": "BRCA1",
      "gene_id": "HGNC:1100",
      "feature_type": { "SoTerm": { "term": "Transcript" } },
      "feature_id": "NM_007294.4",
      "feature_biotype": "Coding",
      "rank": { "ord": 23, "total": 23 },
      "hgvs_t": "c.5536C>A",
      "hgvs_p": "p.Q1846K",
      "tx_pos": { "ord": 5649, "total": 7088 },
      "cds_pos": { "ord": 5536, "total": 5592 },
      "protein_pos": { "ord": 1846, "total": 1864 },
      "distance": -1440,
      "messages": null
    },
    {
      "consequences": ["MissenseVariant"],
      "putative_impact": "Moderate",
      "gene_symbol": "BRCA1",
      "gene_id": "HGNC:1100",
      "feature_type": { "SoTerm": { "term": "Transcript" } },
      "feature_id": "NM_007297.4",
      "feature_biotype": "Coding",
      "rank": { "ord": 22, "total": 22 },
      "hgvs_t": "c.5395C>A",
      "hgvs_p": "p.Q1799K",
      "tx_pos": { "ord": 5589, "total": 7028 },
      "cds_pos": { "ord": 5395, "total": 5451 },
      "protein_pos": { "ord": 1799, "total": 1817 },
      "distance": -1440,
      "messages": null
    },
    {
      "consequences": ["MissenseVariant"],
      "putative_impact": "Moderate",
      "gene_symbol": "BRCA1",
      "gene_id": "HGNC:1100",
      "feature_type": { "SoTerm": { "term": "Transcript" } },
      "feature_id": "NM_007298.3",
      "feature_biotype": "Coding",
      "rank": { "ord": 22, "total": 22 },
      "hgvs_t": "c.2224C>A",
      "hgvs_p": "p.Q742K",
      "tx_pos": { "ord": 2243, "total": 3682 },
      "cds_pos": { "ord": 2224, "total": 2280 },
      "protein_pos": { "ord": 742, "total": 760 },
      "distance": -1440,
      "messages": null
    },
    {
      "consequences": ["MissenseVariant"],
      "putative_impact": "Moderate",
      "gene_symbol": "BRCA1",
      "gene_id": "HGNC:1100",
      "feature_type": { "SoTerm": { "term": "Transcript" } },
      "feature_id": "NM_007300.4",
      "feature_biotype": "Coding",
      "rank": { "ord": 24, "total": 24 },
      "hgvs_t": "c.5599C>A",
      "hgvs_p": "p.Q1867K",
      "tx_pos": { "ord": 5712, "total": 7151 },
      "cds_pos": { "ord": 5599, "total": 5655 },
      "protein_pos": { "ord": 1867, "total": 1885 },
      "distance": -1440,
      "messages": null
    },
    {
      "consequences": ["ThreePrimeUtrVariant"],
      "putative_impact": "Modifier",
      "gene_symbol": "BRCA1",
      "gene_id": "HGNC:1100",
      "feature_type": { "SoTerm": { "term": "Transcript" } },
      "feature_id": "NM_007299.4",
      "feature_biotype": "Coding",
      "rank": { "ord": 22, "total": 22 },
      "hgvs_t": "c.*50C>A",
      "hgvs_p": "p.?",
      "tx_pos": { "ord": 2257, "total": 3696 },
      "cds_pos": { "ord": 50, "total": 2100 },
      "protein_pos": null,
      "distance": -1440,
      "messages": null
    }
  ]
}
