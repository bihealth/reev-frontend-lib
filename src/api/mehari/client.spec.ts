import fs from 'fs'
import path from 'path'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

import { SeqvarImpl } from '../../lib/genomicVars'
import { LinearStrucvarImpl } from '../../lib/genomicVars'
import { setupUrlConfigForTesting, urlConfig } from '../../lib/urlConfig'
import { GenomeBuild } from '../../pbs/mehari/txs'
import { MehariClient } from './client'

/** Fixture Seqvar */
const seqvar = new SeqvarImpl('grch37', '1', 123, 'A', 'G')

/** Fixture with gene trancripts. */
const genesTxsBrca1 = JSON.parse(
  fs.readFileSync(
    path.resolve(__dirname, '../../components/GeneClinvarCard/fixture.genesTxs.BRCA1.37.json'),
    'utf8'
  )
)

/** Fixture with BRCA1 seqvar consequence. */
const seqvarCsqResponseBrca1 = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, './fixture.seqvarCsqResponse.BRCA1.json'), 'utf8')
)

/** Fixture Strucvar affecting BRCA1 */
const strucvar = new LinearStrucvarImpl('DEL', 'grch37', 'chr17', 43044295, 43044297)

/** Fixture with strucvar (BRCA1) consequence */
const strucvarCsqResponseBrca1 = JSON.parse(
  fs.readFileSync(path.resolve(__dirname, './fixture.strucvarCsqResponse.BRCA1.json'), 'utf8')
)

/** Initialize mock for `fetch()`. */
const fetchMocker = createFetchMock(vi)

describe.concurrent('MehariClient.construct()', () => {
  afterEach(() => {
    setupUrlConfigForTesting()
  })

  it('constructs correctly with default base URL', () => {
    // act:
    const client = new MehariClient()

    // assert:
    expect(client).toBeDefined()
  })

  it('constructs correctly with custom base URL', () => {
    // act:
    const client = new MehariClient('http://localhost:8080')

    // assert:
    expect(client).toBeDefined()
  })

  it('throws error if no base URL is configured', () => {
    // arrange:
    urlConfig.baseUrlMehari = undefined

    // (guarded)
    expect(() => new MehariClient(undefined)).toThrow(
      'Configuration error: API base URL not configured'
    )
  })
})

describe.concurrent('MehariClient', () => {
  beforeEach(() => {
    fetchMocker.enableMocks()
    fetchMocker.resetMocks()
  })

  it('properly returns gene transcripts', async () => {
    // arrange:
    fetchMocker.mockResponseOnce(JSON.stringify(genesTxsBrca1))

    // act:
    const client = new MehariClient()
    const result = await client.retrieveGeneTranscripts(
      'HGNC:1100',
      GenomeBuild.GENOME_BUILD_GRCH37
    )

    // assert:
    expect(JSON.stringify(result)).toMatchSnapshot()
  })
})

describe.concurrent('MehariClient/seqvar', () => {
  beforeEach(() => {
    fetchMocker.enableMocks()
    fetchMocker.resetMocks()
  })

  it('fetches TxCsq info correctly', async () => {
    // arrange:
    fetchMocker.mockResponseOnce(JSON.stringify(seqvarCsqResponseBrca1))

    // act:
    const client = new MehariClient()
    const result = await client.retrieveSeqvarsCsq(seqvar, 'HGNC:1100')

    // assert:
    expect(JSON.stringify(result)).toMatchSnapshot()
  })

  it('fetches TxCsq info correctly without HGNC id', async () => {
    // arrange:
    fetchMocker.mockResponseOnce(JSON.stringify(seqvarCsqResponseBrca1))

    // act:
    const client = new MehariClient()
    const result = await client.retrieveSeqvarsCsq(seqvar)

    // assert:
    expect(JSON.stringify(result)).toMatchSnapshot()
  })

  it('fails to fetch variant info with wrong variant', async () => {
    // arrange:
    const seqVarInvalid = new SeqvarImpl('grch37', '1', 123, 'A', 'T')
    fetchMocker.mockResponse((req) => {
      if (req.url.includes('alternative=G')) {
        return Promise.resolve(JSON.stringify(seqvarCsqResponseBrca1))
      }
      return Promise.reject('failed to fetch seqvar')
    })

    // act:
    const client = new MehariClient()
    // (with guard)
    await expect(async () => await client.retrieveSeqvarsCsq(seqVarInvalid)).rejects.toThrow(
      'failed to fetch seqvar'
    )

    // assert:
  })
})

describe.concurrent('MehariClient/strucvar', () => {
  beforeEach(() => {
    fetchMocker.enableMocks()
    fetchMocker.resetMocks()
  })

  it('fetches strucvar info correctly', async () => {
    // arrange:
    fetchMocker.mockResponseOnce(JSON.stringify(strucvarCsqResponseBrca1))

    // act:
    const client = new MehariClient()
    const result = await client.retrieveStrucvarsCsq(strucvar)

    // assert:
    expect(JSON.stringify(result)).toMatchSnapshot()
  })

  it('fails to fetch variant info with wrong variant', async () => {
    // arrange:
    const strucVarInvalid = new LinearStrucvarImpl('DUP', 'grch37', 'chr17', 43044295, 43044297)
    fetchMocker.mockResponse((req) => {
      if (req.url.includes('DEL')) {
        return Promise.resolve(JSON.stringify(strucvarCsqResponseBrca1))
      }
      return Promise.reject('failed to fetch strucvar')
    })

    // act:
    const client = new MehariClient()
    // (with guard)
    await expect(async () => await client.retrieveStrucvarsCsq(strucVarInvalid)).rejects.toThrow(
      'failed to fetch strucvar'
    )

    // assert:
  })
})
