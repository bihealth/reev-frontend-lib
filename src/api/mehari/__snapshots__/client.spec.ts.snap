// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MehariClient/seqvar > fails to fetch variant info with wrong variant 1`] = `"{"version":{"txDb":"0.5.0","mehari":"0.7.0"},"query":{"genomeRelease":"grch37","chromosome":"chr17","position":41197751,"reference":"G","alternative":"T","hgncId":null},"result":[{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007294.4","featureBiotype":"Coding","featureTag":["ManeSelect"],"rank":{"total":23},"hgvsT":"c.5536C>A","hgvsP":"p.Q1846K","txPos":{"ord":5649,"total":7088},"cdsPos":{"ord":5536,"total":5592},"proteinPos":{"ord":1846,"total":1864},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007297.4","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.5395C>A","hgvsP":"p.Q1799K","txPos":{"ord":5589,"total":7028},"cdsPos":{"ord":5395,"total":5451},"proteinPos":{"ord":1799,"total":1817},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007298.3","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.2224C>A","hgvsP":"p.Q742K","txPos":{"ord":2243,"total":3682},"cdsPos":{"ord":2224,"total":2280},"proteinPos":{"ord":742,"total":760},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007300.4","featureBiotype":"Coding","rank":{"total":24},"hgvsT":"c.5599C>A","hgvsP":"p.Q1867K","txPos":{"ord":5712,"total":7151},"cdsPos":{"ord":5599,"total":5655},"proteinPos":{"ord":1867,"total":1885},"distance":-1440},{"consequences":["ThreePrimeUtrVariant"],"putativeImpact":"Modifier","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007299.4","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.*50C>A","hgvsP":"p.?","txPos":{"ord":2257,"total":3696},"cdsPos":{"ord":50,"total":2100},"distance":-1440}]}"`;

exports[`MehariClient/seqvar > fails to fetch variant info with wrong variant 2`] = `"{"version":{"txDb":"0.5.0","mehari":"0.7.0"},"query":{"genomeRelease":"grch37","chromosome":"chr17","position":41197751,"reference":"G","alternative":"T","hgncId":null},"result":[{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007294.4","featureBiotype":"Coding","featureTag":["ManeSelect"],"rank":{"total":23},"hgvsT":"c.5536C>A","hgvsP":"p.Q1846K","txPos":{"ord":5649,"total":7088},"cdsPos":{"ord":5536,"total":5592},"proteinPos":{"ord":1846,"total":1864},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007297.4","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.5395C>A","hgvsP":"p.Q1799K","txPos":{"ord":5589,"total":7028},"cdsPos":{"ord":5395,"total":5451},"proteinPos":{"ord":1799,"total":1817},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007298.3","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.2224C>A","hgvsP":"p.Q742K","txPos":{"ord":2243,"total":3682},"cdsPos":{"ord":2224,"total":2280},"proteinPos":{"ord":742,"total":760},"distance":-1440},{"consequences":["MissenseVariant"],"putativeImpact":"Moderate","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007300.4","featureBiotype":"Coding","rank":{"total":24},"hgvsT":"c.5599C>A","hgvsP":"p.Q1867K","txPos":{"ord":5712,"total":7151},"cdsPos":{"ord":5599,"total":5655},"proteinPos":{"ord":1867,"total":1885},"distance":-1440},{"consequences":["ThreePrimeUtrVariant"],"putativeImpact":"Modifier","geneSymbol":"BRCA1","geneId":"HGNC:1100","featureType":{"SoTerm":{"term":"Transcript"}},"featureId":"NM_007299.4","featureBiotype":"Coding","rank":{"total":22},"hgvsT":"c.*50C>A","hgvsP":"p.?","txPos":{"ord":2257,"total":3696},"cdsPos":{"ord":50,"total":2100},"distance":-1440}]}"`;

exports[`MehariClient/strucvar > fails to fetch variant info with wrong variant 1`] = `"{"version":{"txDb":"0.5.0","mehari":"0.8.0"},"query":{"genomeRelease":"grch37","chromosome":"chr17"},"result":[{"hgncId":"HGNC:16919","transcriptEffects":["upstream_variant"]},{"hgncId":"HGNC:1100","transcriptEffects":["transcript_variant","exon_variant","splice_region_variant","intron_variant","upstream_variant","downstream_variant"]},{"hgncId":"HGNC:18315","transcriptEffects":["transcript_variant","exon_variant","splice_region_variant","intron_variant","upstream_variant","downstream_variant"]},{"hgncId":"HGNC:20691","transcriptEffects":["upstream_variant"]}]}"`;
